library/std/src/sys/wasi/alloc.rs

```diff
@@ -1,42 +1,68 @@
+#![deny(unsafe_op_in_unsafe_fn)]
+
 use crate::alloc::{GlobalAlloc, Layout, System};
 use crate::ptr;
 use crate::sys_common::alloc::{realloc_fallback, MIN_ALIGN};
 
+// SAFETY: All methods implemented follow the contract rules defined
+// in `GlobalAlloc` trait except for `alloc`, which has a an undefined
+// behaviour with `libc::aligned_alloc`, see the FIXME below for more
+// information.
 #[stable(feature = "alloc_system_type", since = "1.28.0")]
 unsafe impl GlobalAlloc for System {
     #[inline]
     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
         if layout.align() <= MIN_ALIGN && layout.align() <= layout.size() {
-            libc::malloc(layout.size()) as *mut u8
+            // SAFETY: `libc::malloc` is guaranteed to be safe, it will allocate
+            // `layout.size()` bytes of memory and return a pointer to it
+            unsafe { libc::malloc(layout.size()) as *mut u8 }
         } else {
-            libc::aligned_alloc(layout.align(), layout.size()) as *mut u8
+            // SAFETY:
+            // FIXME: `libc::aligned_alloc` can fail if `layout.size()` is not
+            // multiple of `layout.align()`, and this constraint does not seem
+            // to be checked
```

Comment:
cc @Amanieu on this FIXME, perhaps you know if we can take this as a guarantee of something (Layout?)

Reply:
Indeed `Layout`'s size may not be a multiple of the alignment. If you need that, use [`pad_to_align`](https://doc.rust-lang.org/nightly/std/alloc/struct.Layout.html#method.pad_to_align).

Reply:
@Amjad50, could you update this to remove the FIXME and call that?

I actually wonder if this is a potential cause for https://github.com/rust-lang/hashbrown/issues/194, but seems somewhat unlikely.

Reply:
Ok